core-primitives/stf-state-handler/src/handle_state.rs

```diff
@@ -22,20 +22,25 @@ use std::sync::SgxRwLockWriteGuard as RwLockWriteGuard;
 use std::sync::RwLockWriteGuard;
 
 use crate::error::Result;
-use itp_types::{ShardIdentifier, H256};
+use itp_types::ShardIdentifier;
 
-/// Facade for handling STF state loading and storing (e.g. from file)
+/// Facade for handling STF state loading and storing (e.g. from file).
 pub trait HandleState {
 	type WriteLockPayload;
 	type StateT;
+	type HashType;
 
-	/// Load the state for a given shard
+	/// Initialize a new shard.
 	///
-	/// Initializes the shard and state if necessary, so this is guaranteed to
-	/// return a state
-	fn load_initialized(&self, shard: &ShardIdentifier) -> Result<Self::StateT>;
+	/// Initializes a default state for the shard and returns its hash.
+	fn initialize_shard(&self, shard: &ShardIdentifier) -> Result<Self::HashType>;
 
-	/// Load the state in order to mutate it
+	/// Load the state for a given shard.
+	///
+	/// Requires the shard to exist and be initialized, otherwise returns an error.
+	fn load(&self, shard: &ShardIdentifier) -> Result<Self::StateT>;
```

Comment:
The initialization and loading are two separate methods. Loading is now a read-only method, but it requires the shard to be initialized explicitly using `initialize_shard`.